src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpUseRecursivePatternsCodeRefactoringProvider)), Shared]
+    internal sealed partial class CSharpUseRecursivePatternsCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var textSpan = context.Span;
+            if (textSpan.Length > 0)
+            {
+                return;
+            }
+
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = GetOutermostExpression(root.FindToken(textSpan.Start).Parent);
+            if (node is null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            // First, we make a tree of comparisons and pattern-matches from source.
+            var analyzedNode = Analyzer.Analyze(node, semanticModel);
+            if (analyzedNode is null)
+            {
+                return;
+            }
+
+            // Then, we try to combine common expressions to rewrite matches as a recursive-pattern.
+            var reducedNode = Reducer.Reduce(analyzedNode, out var isNonTrivial);
+            if (reducedNode is null || !isNonTrivial)
```

Comment:
why do you need isNonTrivial?  if it's just going to be used to bail out here, coudln't it just be used internally to decide to just return null?